### PR TITLE
Reject faked stake/vote accounts in stake mgmt.

### PR DESCRIPTION
**Pull request by ryoqun**
#### Problem

There is no `account.owner` check for referenced stake/vote accounts when delegating/splitting/merging stake accounts.

Because these referenced accounts are read-only, the runtime doesn't enforce the `account.owner` check. Users can pass any accounts of any owner as long as it deserializes into `StakeState` or `VoteState`, so we must check `account.owner` by ourselves.

This time I checked this trait for possible similar issues, but I found other uses of `KeyedAccount` without `owner` check is legitimate: https://github.com/solana-labs/solana/blob/6c5fb329b292c1abf7aaf62cf5558f5a9a1ca98b/programs/stake/src/stake_state.rs#L53-L106

#### Summary of Changes

Add guards and outright reject such transactions.

Also, this is feature-gated under `stake_program_v2`, implicitly.

Fixes #

### CODE REVIEW

**Comment by ryoqun** [comment]
@CriesofCarrots Could you review this before #13461 ?

@rwalker-com Hi, could you review this?

File: programs/stake/src/stake_instruction.rs [hunk 1]
```diff
@@ -552,6 +561,14 @@ mod tests {
...
                 } else if sysvar::rent::check_id(&meta.pubkey) {
                     account::create_account(&Rent::default(), 1)
+                } else if meta.pubkey == bad_stake_pubkey() {
+                    let mut account = Account::default();
+                    account.owner = id();
+                    account
+                } else if meta.pubkey == bad_vote_pubkey() {
+                    let mut account = Account::default();
+                    account.owner = solana_vote_program::id();
+                    account
```
Comment (CriesofCarrots) [inline programs/stake/src/stake_instruction.rs:571]: These are for test cases where the owner should be correct but the account data can't be deserialized as StakeState/VoteState respectively, correct?
I found that a little confusing in the context of the rest of the PR, which deals with valid accounts with bad owners.
I think renaming the pubkey functions a little would help... I'll suggest above

File: programs/stake/src/stake_instruction.rs [hunk 2]
```diff
@@ -532,11 +532,20 @@ mod tests {
...
         RefCell::new(Account::default())
     }
 
+    fn bad_stake_pubkey() -> Pubkey {
+        Pubkey::from_str("BadStake11111111111111111111111111111111111").unwrap()
+    }
+
+    fn bad_vote_pubkey() -> Pubkey {
+        Pubkey::from_str("BadVote111111111111111111111111111111111111").unwrap()
+    }
```
Comment (CriesofCarrots) [inline programs/stake/src/stake_instruction.rs:547]:
```suggestion
    fn invalid_stake_state_pubkey() -> Pubkey {
        Pubkey::from_str("BadStake11111111111111111111111111111111111").unwrap()
    }

    fn invalid_vote_state_pubkey() -> Pubkey {
        Pubkey::from_str("BadVote111111111111111111111111111111111111").unwrap()
    }
```

**Review by CriesofCarrots — APPROVED** [review]
Fix and process tests lgtm!
One nit to make stake_instruction test changes more clear.

File: programs/stake/src/stake_state.rs [hunk 3]
```diff
@@ -821,6 +821,10 @@ impl<'a> StakeAccount for KeyedAccount<'a> {
         config: &Config,
         signers: &HashSet<Pubkey>,
     ) -> Result<(), InstructionError> {
+        if vote_account.owner()? != solana_vote_program::id() {
+            return Err(InstructionError::IncorrectProgramId);
+        }
```
Comment (ryoqun) [inline programs/stake/src/stake_state.rs:826]: @jackcmay just making sure; it seems that `IncorrectProgramId` isn't used much; but I think this is a perfect usecase of `IncorrectProgramId` for these kinds of checks, I guess.

File: programs/stake/src/stake_instruction.rs (hunk 1 above)
Comment (ryoqun) [reply programs/stake/src/stake_instruction.rs]: thanks for the suggestion!: done: https://github.com/solana-labs/solana/pull/13615/commits/933dcfd465775638f63cc323227d0b40cd6d55d6

**Review by rwalker-com — APPROVED** [review]
lgtm

File: programs/stake/src/stake_state.rs (hunk 3 above)
Comment (jackcmay) [reply programs/stake/src/stake_state.rs:826]: The description of this error is doesn't make much sense but the way you propose to use it here makes as much sense as any ;-)

**Comment by ryoqun** [comment]
@CriesofCarrots @rob-solana Thanks for reviewing! I'm going to merge this shortly after now. I'm doing last-minute local test this.

File: programs/stake/src/stake_state.rs (hunk 3 above)
Comment (CriesofCarrots) [reply programs/stake/src/stake_state.rs:826]: Fwiw, this seems perfectly consistent with how the error is used in the Vest program

Comment (ryoqun) [reply programs/stake/src/stake_state.rs:826]: yeah, I copied from that. :)

**Comment by ryoqun** [comment]
> buildkite/solana/downstream-projects — Failed (exit status 101) 

Merging this despite the quoted build failure; I'm aware this is already being addressed and the failure isn't related to this pr.

**Pull request merged**